File: annotate/db/annotate.scala
import scala.math.random
import spark._
import SparkContext._
import scala.util.matching.Regex
import scala.collection.mutable.ListBuffer
import scala.io._

object annotate {

  def map1(line:String){

    val word:String = (line.trim().toLowerCase())
    word
  }

  def map3(line:String,names:spark.broadcast.Broadcast[List[String]],annotate_text:String):String = {
    val words = line.split("\t")
    var l = "<-->"
    var word1 = ""
    var word2 = ""
    var flag = 0
    for(i <- 0 to (names.value.length - 1)){
     // println("words(1)->"+(" " + words(1).trim.toLowerCase +" ")+"|")
     // println("words(3)->"+(" " + words(3).trim.toLowerCase +" ")+"|")
     // println("name->"+(" "+names.value(i).trim.toLowerCase+" ")+"|")
      if((" " + words(1).trim.toLowerCase +" ") contains (" "+names.value(i).trim.toLowerCase+" ")) {
        l = line
        if(words(1).trim.toLowerCase contains "</type>"){
          val pat = "<type=(\\w+)>.+</type>".r
          var type1:String = ""
          (pat findAllIn words(1)).matchData foreach{ m =>
            type1 = m.group(1)
          }
          val annotated_line = (words(1)).replaceAll("<type="+type1+">","<type="+annotate_text.toUpperCase+"|"+type1.toUpperCase+">")
          words(1) = annotated_line.trim
          flag = 1
        }else{
          val annotated_line = (" " + words(1).trim.toLowerCase +" ").replaceAll(" "+names.value(i).trim.toLowerCase+" "," <type="+annotate_text.toUpperCase+"> "+names.value(i)+" </type> ")
          words(1) = annotated_line.trim
          flag = 1
        }
      }
    }


    for(i <- 0 to (names.value.length - 1)){
      if((" " + words(3).trim.toLowerCase +" ") contains (" "+names.value(i).trim.toLowerCase+" ")){
        l = line
        if(words(3).trim.toLowerCase contains "</type>"){
          val pat = "<type=(\\w+)>.+</type>".r
          var type1:String = ""
          (pat findAllIn words(3)).matchData foreach{ m =>
            type1 = m.group(1)
          }
          val annotated_line = (words(3)).replaceAll("<type="+type1+">","<type="+annotate_text.toUpperCase+"|"+type1.toUpperCase+">")
          words(3) = annotated_line.trim
          flag = 1
        }else{
          val annotated_line = (" " + words(3).trim.toLowerCase +" ").replaceAll(" "+names.value(i).trim.toLowerCase+" "," <type="+annotate_text.toUpperCase+"> "+names.value(i)+" </type> ")
          words(3) = annotated_line.trim
          flag = 1
        }
      }
    }


    if(flag == 0){
      return line
      }else{
        return words(0)+"\t"+words(1)+"\t"+words(2)+"\t"+words(3)
      }
    
  }

  

  def main(args: Array[String]) {
    if (args.length == 0) {
      System.err.println("Usage: annotate annotate_input_file annotate_use_file annotate_text <master>")
      System.exit(1)
    }
    val spark = new SparkContext(args(4), "annotate",  System.getenv("SPARK_HOME"), List(System.getenv("SPARK_TEST")))
    val store = Source.fromFile(args(0))
    val data = spark.textFile(args(1))
    val annotate_text = args(2)

    var names:List[String] = Nil

    //println(args(0))
    //println(args(1))
    //println(args(2))
    //println(args(3))
    store.getLines.foreach((line) => {names = names:::List(line.trim)})

    val names_data = spark.broadcast(names)

    val data_set = data.map(line => map3(line,names_data,annotate_text))
    //val clean_data_set = data_set.filter(line => if(line.contains("<-->")) false else true).cache()

    //val st = store.flatMap(line =>  map1(line))

    //val dt = data.flatMap(line => map2(line,st))
    data_set.saveAsTextFile(args(3))
    System.exit(0)
  }
}

File: annotate/db/annotate.py
import re
import sys

file_store = open(sys.argv[1],'r')
names = []
names1 = []
for name in file_store:
	names.append(name[:len(name)-1].lower())
print len(names)

names1 = names[:]
print len(names1)
#names = ["shahrukh khan"]


data = open(sys.argv[2],'r')

annotate_text = sys.argv[3].upper()
type1 = ""
type2 = ""
word1 = ""
word2 = ""



def annotate_arg1(names,words,annsotate_text):
	done1 = 0
	done2 = 0
	for name in names:
		#print names
			
		name = name.strip().lower()

		m = re.search("<type=(.*)>(.*)</type>",words[1].lower())
		if m and done1 == 0:
			#print "typed"
			word1 = m.group(2)
			type1 = m.group(1)
			#print "WORD=",word1
			if re.search(word1.lower().strip(),name):
				words[1] = "<type="+type1+"|"+annotate_text+">"+word1+"</type>"
				#print words[1]
				done1 = 1
		
		elif done1 == 0:	
			#print name,words[1]
			name = name.replace("?"," ")
			name = name.replace("*"," ")
			name = name .replace(".","\.")
			name = name.replace("+"," ")
			name = name.replace("["," ")
			name = name.replace("]"," ")
			if not re.search("[\(]|[\)]" , name):
				m = re.search(name,words[1].lower())
				if m:
					#print "untyped:arg1"
					words[1] = "<type="+annotate_text+">"+name+"</type>"
					done1 = 1


		m = re.search("<type=(.*)>(.*)</type>",words[3].lower())
		if m and done2 == 0:
			
			word1 = m.group(2)
			type1 = m.group(1)
			#print "WORD=",word1
			if len(name)> 0 and re.search(word1.lower().strip(),name):
				#print "typed:arg2"
				words[3] = "<type="+type1+"|"+annotate_text+">"+word1+"</type>"
				#print words[1]
				done2 == 1
		
		elif done2 == 0:	
			name = name.replace("?"," ")
			name = name .replace(".","\.")
			name = name.replace("*"," ")
			name = name.replace("+"," ")
			name = name.replace("["," ")
			name = name.replace("]"," ")
			if len(name) > 0 and not re.search("[\(]|[\)]" , name):
				#print "arg2-name",name
				m = re.match(name,words[3].lower())
				if m:
					#print "untyped:arg2"
					words[3] = "<type="+annotate_text+">"+name+"</type>"
					done2 = 1
		if done1 == 1 and done2 == 1:
			return 1

	if done1 == 1 or done2 == 1:
		return 1
	return 0


def annotate_arg2(names1,words,annotate_text):
	for name in names:
		#print names
			
		name = name.strip().lower()


		m = re.search("<type=(.*)>(.*)</type>",words[3].lower())
		if m:
			print "typed:arg2"
			word1 = m.group(2)
			type1 = m.group(1)
			#print "WORD=",word1
			if re.search(word1.lower().strip(),name):
				words[3] = "<type="+type1+"|"+annotate_text+">"+word1+"</type>"
				#print words[1]
				return 1
		
		else:	
			
			if not re.search("\(" , name) and len(name) > 0:
				m = re.search(name,words[3].lower())
				if m:
					print name,words[3]
					print "untyped:arg2"
					words[3] = "<type="+annotate_text+">"+name+"</type>"
					return 1
	return 0

for line in data:
	
	words = line.split("\t")
	print words
	flag = 0
	flag = annotate_arg1(names,words,annotate_text)	
	words[3] = words[3].replace('\n','')

	#flag = annotate_arg2(names,words,annotate_text)	

	if flag ==  1:
		print "\t".join(words)
	#else:
	#	print line